Source/common/SNTConfigurator.m

```diff
@@ -322,70 +340,79 @@ - (NSString *)machineID {
+  if (!self.configData[kSyncBaseURLKey]) return;
+
+  // Load the last known sync state
+  if (![[NSFileManager defaultManager] fileExistsAtPath:kSyncStateFilePath]) return;
 
   NSError *error;
-  NSData *readData = [NSData dataWithContentsOfFile:self.configFilePath
+  NSData *readData = [NSData dataWithContentsOfFile:kSyncStateFilePath
                                             options:NSDataReadingMappedIfSafe
                                               error:&error];
-  if (error) {
-    LOGE(@"Could not read configuration file: %@, replacing.", [error localizedDescription]);
-    [self saveConfigToDisk];
+  if (!readData) {
+    LOGE(@"Could not read sync state file: %@, replacing.", [error localizedDescription]);
     return;
   }
 
-  NSMutableDictionary *configData =
+  NSMutableDictionary *syncState =
       [NSPropertyListSerialization propertyListWithData:readData
                                                 options:NSPropertyListMutableContainers
                                                  format:NULL
                                                   error:&error];
-  if (error) {
-    LOGE(@"Could not parse configuration file: %@, replacing.", [error localizedDescription]);
-    [self saveConfigToDisk];
+  if (!syncState) {
+    LOGE(@"Could not parse sync state file: %@, replacing.", [error localizedDescription]);
     return;
   }
 
-  if (self.syncBaseURL) {
-    // Ensure no-one is trying to change protected keys behind our back.
-    BOOL changed = NO;
-    if (geteuid() == 0) {
-      for (NSString *key in self.protectedKeys) {
-        if (((self.configData[key] && !configData[key]) ||
-             (!self.configData[key] && configData[key]) ||
-             (self.configData[key] && ![self.configData[key] isEqual:configData[key]]))) {
-          if (self.configData[key]) {
-            configData[key] = self.configData[key];
-          } else {
-            [configData removeObjectForKey:key];
-          }
-          changed = YES;
-          LOGI(@"Ignoring changed configuration key: %@", key);
-        }
-      }
-    }
-    self.configData = configData;
-    if (changed) [self saveConfigToDisk];
-  } else {
-    self.configData = configData;
+  // Overwrite or add the sync state to the running config
+  for (NSString *key in [self syncServerKeys]) {
+    self.configData[key] = syncState[key];
   }
 }
 
 #pragma mark Private
 
 ///
-///  Saves the current @c self.configData to disk.
+///  Saves the current effective syncState to disk.
 ///
-- (void)saveConfigToDisk {
+- (void)saveSyncStateToDisk {
   if (geteuid() != 0) return;
-  [self.configData writeToFile:self.configFilePath atomically:YES];
+  NSMutableDictionary *syncState =
+      [NSMutableDictionary dictionaryWithCapacity:[self syncServerKeys].count];
+  for (NSString *key in [self syncServerKeys]) {
+    syncState[key] = self.configData[key];
+  }
+  [syncState writeToFile:kSyncStateFilePath atomically:YES];
```

Comment:
Probably want to ensure here that the permissions aren't broken.

Reply:
It's probably worth watching the sync state file for changes and re-writing the file if it changes behind our back. santad should be the only process writing to the file.